src/verso-manual/VersoManual.lean

```diff
@@ -328,17 +324,28 @@ def emitTeX (logError : String → IO Unit) (config : Config) (text : Part Manua
       | continue
     packages := packages.insertMany d.usePackages
     preambleItems := preambleItems.insertMany d.preamble
-  withFile (dir.join "main.tex") .write fun h => do
-    if config.verbose then
-      IO.println s!"Saving {dir.join "main.tex"}"
-    h.putStrLn (preamble text.titleString authors date packages.toList preambleItems.toList)
+  let putStrLn (line : String) : StateT String Id Unit := do
+    modify (fun s => s ++ line ++ "\n")
+  let tex : StateT String Id Unit := do
+    putStrLn (preamble text.titleString authors date packages.toList preambleItems.toList)
     if frontMatter.size > 0 then
-      h.putStrLn "\\chapter*{Introduction}"
+      putStrLn "\\chapter*{Introduction}"
     for b in frontMatter do
-      h.putStrLn b.asString
+      putStrLn b.asString
     for c in chapters do
-      h.putStrLn c.asString
-    h.putStrLn postamble
+      putStrLn c.asString
+    putStrLn postamble
+  pure (StateT.run tex "").2
+
+open IO.FS in
+def emitTeX (logError : String → IO Unit) (config : Config) (text : Part Manual) : ReaderT ExtensionImpls IO Unit := do
+  let dir := config.destination.join "tex"
+  ensureDir dir
+  let tex ← makeTeX logError config text
```

Comment:
Rather than making this a string, what if we passed an `IO.FS.Stream` here, potentially via a Reader? That can be hooked up to a file handle, but it can also be set up to log the output. That way, we don't commit ourselves to creating a string in memory for all TeX output, but we can still test the output by doing so.